examples/hitl/rearrange_v2/object_state_manipulator.py

```diff
@@ -0,0 +1,172 @@
+        Returns None if no action is available to do this state change.
+        """
+        if type(target_value) == bool:
+            return self.get_boolean_action(state_name, target_value)
+        else:
+            # Unsupported action type.
+            return None
+
+    def can_execute_action(
+        self, state_name: str, state_value: Any, object_handle: str
+    ) -> Tuple[bool, str]:
+        """
+        Test whether an action can be executed on the specified object.
+        Returns whether the action can be executed, and an error message if it can't.
+        """
+        action = self.get_action(state_name, state_value)
+        if action is not None:
+            return action.can_modify_state_impl(
+                self._sim,
+                self._agent_index,
+                object_handle,
+                state_name,
+                self._maximum_distance,
+                self._world._metadata_interface,
+            )
+        return (False, f"Undefined action: '{state_name} -> {state_value}'.")
+
+    def try_execute_action(
+        self, state_name: str, state_value: Any, object_handle: str
+    ) -> Tuple[bool, str]:
+        """
+        Try executing an action on the specified object.
+        Returns whether the action was successful, and an error message if it wasn't.
+        """
+        success, error_message = self.can_execute_action(
+            state_name, state_value, object_handle
+        )
+        if success:
+            self.set_object_state(object_handle, state_name, state_value)
+        return (success, error_message)
+
+    def get_all_available_boolean_actions(
+        self, object_handle: str
+    ) -> List[BooleanAction]:
+        """
+        Return all boolean actions available for the specified object.
+        """
+        actions: List[BooleanAction] = []
+        world = self._world
+        states = world.get_states_for_object_handle(object_handle)
+        for state in states:
+            spec = state.state_spec
+            if spec.type == bool:
+                name = spec.name
+                value = cast(bool, state.value)
+                target_value = not value
+                action = self.get_boolean_action(name, target_value)
+                action_available = action is not None
+                action_enabled = False
+                error = None
+                if action_available:
+                    action_enabled, error = self.can_execute_action(
+                        name, target_value, object_handle
+                    )
+                actions.append(
+                    BooleanAction(
+                        state_spec=spec,
+                        current_value=state.value,
+                        target_value=target_value,
+                        available=action_available,
+                        enabled=action_enabled,
+                        error=error,
+                    )
```

Comment:
So you cache the actions available in the world for all objects and then pass these to the UI to populate with based on user's clicks?

Reply:
Correct!